File: backend/app/api/v1/schemas/recommend.py
"""Pydantic schemas for recommendation API."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class FertilizerRecommendationResponse(BaseModel):
    """A scored fertilizer recommendation."""

    fertilizer_id: str
    fertilizer_name: str
    formulation: str
    score: float
    crop_need_score: float
    soil_deficiency_score: float
    cost_efficiency_score: float
    compatibility_score: float
    recommended_rate_kg_per_ha: int
    cost_per_ha_mmk: int
    reasoning: str


class CropRecommendationResponse(BaseModel):
    """Crop + fertilizer pairing recommendation."""

    crop_id: str
    crop_name: str
    crop_name_mm: str
    portfolio_weight: float
    expected_income_per_ha: float
    fertilizers: list[FertilizerRecommendationResponse]


class SoilProfileResponse(BaseModel):
    """Soil profile for a township."""

    township_id: str
    ph_h2o: float
    nitrogen_g_per_kg: float
    soc_g_per_kg: float
    clay_pct: int
    sand_pct: int
    silt_pct: int
    cec_cmol_per_kg: float
    texture_class: str
    fertility_rating: str


class ConfidenceMetrics(BaseModel):
    """Monte Carlo confidence metrics for a recommendation."""

    num_simulations: int
    mean_income: float
    median_income: float
    percentile_5: float
    percentile_95: float
    prob_catastrophic_loss: float
    success_probability: float


class RecommendRequest(BaseModel):
    """Request for crop-fertilizer recommendations."""

    township_ids: list[str] = Field(min_length=1, max_length=10)
    crop_ids: list[str] = Field(min_length=2, max_length=10)
    risk_tolerance: float = Field(default=0.5, ge=0.0, le=1.0)
    season: Literal["monsoon", "dry"] = "dry"
    top_fertilizers: int = Field(default=3, ge=1, le=8)


class TownshipRecommendation(BaseModel):
    """Full recommendation for a single township."""

    township_id: str
    township_name: str
    season: str
    soil: SoilProfileResponse | None
    crops: list[CropRecommendationResponse]
    confidence: ConfidenceMetrics | None
    expected_income_per_ha: float
    risk_reduction_pct: float
    ai_advisory: str | None = None
    ai_advisory_mm: str | None = None


class RecommendResponse(BaseModel):
    """Response with recommendations for one or more townships."""

    recommendations: list[TownshipRecommendation]
    total_townships: int


class DemoROIRequest(BaseModel):
    """Request for demo crop ROI calculation."""

    township_id: str
    crop_id: str
    area_hectares: float = Field(default=1.0, ge=0.1, le=100.0)
    season: Literal["monsoon", "dry"] = "dry"


class DemoROIResponse(BaseModel):
    """ROI calculation for a demo crop scenario."""

    township_id: str
    township_name: str
    crop_id: str
    crop_name: str
    area_hectares: float
    season: str
    # Costs
    fertilizer_cost_mmk: int
    seed_cost_mmk: int
    total_input_cost_mmk: int
    # Returns
    expected_revenue_mmk: int
    expected_profit_mmk: int
    # Risk
    success_probability: float
    catastrophic_loss_probability: float
    reimbursement_exposure_mmk: int
    # Alternatives
    recommended_fertilizer: FertilizerRecommendationResponse | None
    soil: SoilProfileResponse | None

File: frontend/src/types/recommend.ts
/** Scored fertilizer recommendation for a crop-soil combination. */
export interface FertilizerRecommendation {
  fertilizer_id: string;
  fertilizer_name: string;
  formulation: string;
  score: number;
  crop_need_score: number;
  soil_deficiency_score: number;
  cost_efficiency_score: number;
  compatibility_score: number;
  recommended_rate_kg_per_ha: number;
  cost_per_ha_mmk: number;
  reasoning: string;
}

/** Crop + fertilizer pairing recommendation. */
export interface CropRecommendation {
  crop_id: string;
  crop_name: string;
  crop_name_mm: string;
  portfolio_weight: number;
  expected_income_per_ha: number;
  fertilizers: FertilizerRecommendation[];
}

/** Soil profile for a township. */
export interface SoilProfile {
  township_id: string;
  ph_h2o: number;
  nitrogen_g_per_kg: number;
  soc_g_per_kg: number;
  clay_pct: number;
  sand_pct: number;
  silt_pct: number;
  cec_cmol_per_kg: number;
  texture_class: string;
  fertility_rating: string;
}

/** Monte Carlo confidence metrics. */
export interface ConfidenceMetrics {
  num_simulations: number;
  mean_income: number;
  median_income: number;
  percentile_5: number;
  percentile_95: number;
  prob_catastrophic_loss: number;
  success_probability: number;
}

/** Full recommendation for a single township. */
export interface TownshipRecommendation {
  township_id: string;
  township_name: string;
  season: string;
  soil: SoilProfile | null;
  crops: CropRecommendation[];
  confidence: ConfidenceMetrics | null;
  expected_income_per_ha: number;
  risk_reduction_pct: number;
  ai_advisory: string | null;
  ai_advisory_mm: string | null;
}

/** Request body for recommendations. */
export interface RecommendRequest {
  township_ids: string[];
  crop_ids: string[];
  risk_tolerance?: number;
  season?: "monsoon" | "dry";
  top_fertilizers?: number;
}

/** Response with recommendations for one or more townships. */
export interface RecommendResponse {
  recommendations: TownshipRecommendation[];
  total_townships: number;
}

/** Request body for demo ROI calculation. */
export interface DemoROIRequest {
  township_id: string;
  crop_id: string;
  area_hectares?: number;
  season?: "monsoon" | "dry";
}

/** ROI calculation response for a demo crop scenario. */
export interface DemoROIResponse {
  township_id: string;
  township_name: string;
  crop_id: string;
  crop_name: string;
  area_hectares: number;
  season: string;
  fertilizer_cost_mmk: number;
  seed_cost_mmk: number;
  total_input_cost_mmk: number;
  expected_revenue_mmk: number;
  expected_profit_mmk: number;
  success_probability: number;
  catastrophic_loss_probability: number;
  reimbursement_exposure_mmk: number;
  recommended_fertilizer: FertilizerRecommendation | null;
  soil: SoilProfile | null;
}
